### ISSUE TITLE
speedup of single-cell-bulk model: lookup tables for amplification distributions

### ISSUE THREAD

**Issue by dlaehnemann** [PROSIC/libprosic#27]
@johanneskoester, do you think transforming the function `prob_rho()` in the single cell bulk model into a static pre-defined lookup table would yield performance improvements? My general idea would be to implement it as a three-layer array of arrays, with the outer-most level representing alt allele frequencies `0.0`, `0.5` and `1.0` (here a clear mapping of indexes needs to be in the comments), the second level representing the total coverage at a site (with indexes corresponding to total coverage) and the inner-most level representing the number of reads with an alternative allele (again with indexes corresponding to number of reads). Do you know if this will make a difference in Rust?

https://github.com/PROSIC/libprosic/blob/7dfda4ef909015a3fb31557bfc0ab7a2b29b7386/src/model/priors/single_cell_bulk.rs#L42-115

**Comment by johanneskoester** [comment]
This is hard to tell without a benchmarks. But in general, yes, I think three memory lookups might be faster than the current code. It is worth a try at least. Maybe use `VecMap` for this. It adds some convenience. If this can wait some days: I am about to do a huge merge into the master branch which touches some parts of the bulk model as well (at least the tests).

**Comment by dlaehnemann** [comment]
Yes, I guess it will save me some hassle to first wait for that merge before starting a new branch and having to rebase it later.

Some background: Even though this will probably not be necessary in using ProSolo later (to that extent), benchmarking ProSolo on a lot of negative sites (linearly) blows up the runtime (even with massive parallelisation by genomic region, it takes more than a day per sample), so easy gains in runtime per site would be welcome. But the current version works, I am running it on a sample, and I should get results within days -- so it is not a show-stopper.

As for benchmarks, do you have any recommendations which crate(s) to use? Here's what I found:
* in unstable: <https://doc.rust-lang.org/nightly/unstable-book/library-features/test.html>
* separate crate: <https://docs.rs/bencher/0.1.5/bencher/>
* separate crate: <https://doc.rust-lang.org/time/time/index.html>

**Comment by dlaehnemann** [comment]
Hmm, ProSolo does seem to take too long per sample with so many sites. So I should probably go and benchmark to hopefully find candidate functions for optimiziation. Any estimate of when you will merge your changes?

I'll start with a bit of benchmarking, which I can just do on a local branch -- but for the optimizations, it would be nice to have the other stuff merged before.

**Comment by johanneskoester** [comment]
Weekend is close, but I hope to do that early next week.

**Comment by dlaehnemann** [comment]
Instead of hard-coding all the values, @johanneskoester suggested to keep the current formulae and just calculate the tables once upon instantiation of the respective sample model. This code will be much easier to change and to abstract into further different types of amplification models.

**Comment by dlaehnemann** [comment]
And @johanneskoester suggested `nd-array`s for storing the lookup tables.

**Comment by dlaehnemann** [comment]
This has been addressed by caching the function `prob_rho()` in PR #46
